### PR TITLE
xkafka: Sequential & Async Consumer changes 

### LINKED ISSUE #25: xkafka: Sequential & Async Consumer Patterns
Latest version - https://github.com/gojekfarm/xtools/issues/25#issuecomment-1805073951

# v0
This is a proposal to define the behavior of `xkafka.Consumer` for sequential and async consumption use cases.

The processing mode is determined by the `xkafka.Concurrency` option.

## Sequential Processing

Sequential processing is the default mode. It is same as `xkafka.Concurrency(1)`. By default, the consumer is initialized with settings as per https://github.com/confluentinc/librdkafka/wiki/Consumer-offset-management. 

The offset can also be moved manually by enabling `xkafka.ManualOffset` option. In this case, the consumer will store the offset after the message is processed and the `message.Status` is `Success` or `Skip`. The stored offsets will be automatically committed, unless the `enable.auto.commit` option is set to `false`. 

```go
handler := xkafka.HandlerFunc(func(ctx context.Context, message *xkafka.Message) error {
    // process message
    return nil
})

// default sequential processing
consumer := xkafka.NewConsumer(
    "default",
    handler,
    xkafka.Brokers{"localhost:9092"},
    xkafka.Topics{"test"},
)

// manual commit handler
handler := xkafka.HandlerFunc(func(ctx context.Context, message *xkafka.Message) error {
    // ... process message

    message.AckSuccess()

    return nil
})

consumer := xkafka.NewConsumer(
    "manual-commit",
    handler,
    xkafka.Brokers{"localhost:9092"},
    xkafka.Topics{"test"},
    xkafka.ManualOffset(true),
)
```

## Async Processing

Async processing is enabled by setting `xkafka.Concurrency` to a value greater than 1. The consumer will use a pool of Go routines to process messages concurrently.

### Manual Offset 

When `ManualOffset` is enabled in async processing mode, the consumer will process messages concurrently using a pool, but will store the offset sequentially.

But, this approach comes with a caveat. When an earlier message fails after later messages have already been processed, the consumer will not move the offset. Adding a retry middleware or a dead letter queue can help solve this problem.

## Error Handling

Error handling can be done by:
- Within the handler implementation
- Using a error handling middleware or a retry middleware 
- through the catch all `xkafka.ErrorHandler` option 

`xkafka.ErrorHandler` is called for each error either from the handler or internally from kafka. Unhandled errors will cause the consumer to stop processing and panic. 

## Queues

Both sequential and async processing modes require some form of a queue to guarantee at-least-once processing in case of errors, crashes, or restarts.

The queue can be a local on-disk queue (BadgerDB, SQLite, etc.) or a distributed queue (Redis, RabbitMQ, etc.).

Consumers can be configured to use a queue by setting `xkafka.Queue` option. The queue must implement `xkafka.Queuer` interface.

```go
type Queuer interface {
    Put(ctx context.Context, message *Message) error
    Get(ctx context.Context, id string) (*Message, error)
    Delete(ctx context.Context, id string) error
    All(ctx context.Context, cb func(message *Message) error) error
}

queue := xkafkaqueue.NewSQLiteQueue("default", "/tmp/xkafka.db")

consumer := xkafka.NewConsumer(
    "default",
    handler,
    xkafka.Brokers{"localhost:9092"},
    xkafka.Topics{"test"},
    xkafka.Queue(queue),
)
```
~

### PR DESCRIPTION
**Pull request by sonnes**
Fixes #25 

The processing mode is determined by the `xkafka.Concurrency` option. By default, the consumer is initialized with `enable.auto.offset.store=false`. The offset is "stored" after the message is processed. The offset is "committed" based on the `enable.auto.commit` & `auto.commit.interval.ms` options.

It is important to understand the difference between "store" and "commit". The offset is "stored" in the consumer's memory and is "committed" to Kafka. The offset is "stored" after the message is processed and the `message.Status` is `Success` or `Skip`. The stored offsets will be automatically committed, unless the `enable.auto.commit` option is set to `false`.

## Error Handling

By default, `xkafka.Consumer` will stop processing, commit last stored offset, and exit if there is a Kafka error or if the handler returns an error.

Errors can be handled by using one or more of the following options:

- Within the handler implementation
- Using error handling & retry middlewares
- through the catch all `xkafka.ErrorHandler` option

`xkafka.ErrorHandler` is called for every error that is not handled by the handler or the middlewares. It is also called for errors returned by underlying Kafka client.

## Sequential Processing

Sequential processing is the default mode. It is same as `xkafka.Concurrency(1)`.

## Async Processing

Async processing is enabled by setting `xkafka.Concurrency` to a value greater than 1. The consumer will use a pool of Go routines to process messages concurrently.

## Manual Commit

By default, the consumer will automatically commit the offset based on the `enable.auto.commit` & `auto.commit.interval.ms` options. The offsets are committed asynchronously in the background.

The consumer can be configured to commit the offset manually by setting `xkafka.EnableManualCommit` option to `true`. When `ManualCommit` is enabled, the consumer will synchronously commit the offset after each message is processed.

NOTE: Enabling `ManualCommit` will add an overhead to each message. It is recommended to use `ManualCommit` only when necessary.

### CODE REVIEW

File: xkafka/consumer.go [hunk 1]
```diff
@@ -77,28 +84,33 @@ func (c *Consumer) Start(ctx context.Context) error {
...
-	st := stream.New().WithMaxGoroutines(c.config.concurrency)
+	if c.config.concurrency > 1 {
+		return c.runAsync(ctx)
+	}
+
+	return c.runSequential(ctx)
+}
+
+func (c *Consumer) runSequential(ctx context.Context) error {
 	errChan := make(chan error, 1)
```
Comment (ajatprabha) [inline xkafka/consumer.go:95]: Can you eliminate the use of `errChan`, I feel that the code will be simpler without it.

File: xkafka/consumer.go [hunk 2]
```diff
@@ -111,23 +123,100 @@ func (c *Consumer) Start(ctx context.Context) error {
...
+			Offset:    kafka.Offset(msg.Offset + 1),
+		},
+	})
+	if err != nil {
+		return err
+	}
 
-	if ferr := c.config.errorHandler(err); ferr != nil {
-		errChan <- ferr
+	if c.config.manualCommit {
```
Comment (ajatprabha) [inline xkafka/consumer.go:212]: This condition may always force a strict commit behaviour. Say I want to commit after every 10th message(or offset), how will this allow that? If that is out of scope for now, is there room for improvement in this area later on?

**Review by ajatprabha — CHANGES_REQUESTED** [review]
Quick review done, left some comments, will do a thorough one soon.

File: xkafka/consumer.go [hunk 3]
```diff
@@ -111,23 +123,100 @@ func (c *Consumer) Start(ctx context.Context) error {
...
 
-func (c *Consumer) runHandler(ctx context.Context, handler Handler, km *kafka.Message, errChan chan error) {
-	msg := newMessage(c.name, km)
+func (c *Consumer) storeMessage(msg *Message) error {
+	if msg.Status != Success && msg.Status != Skip {
+		return nil
+	}
 
-	err := handler.Handle(ctx, msg)
+	_, err := c.kafka.StoreOffsets([]kafka.TopicPartition{
```
Comment (ajatprabha) [inline xkafka/consumer.go:201]: Offset storage is quite unintuitive to reason about, can we add some documentation with few examples around this?
Explaining its behaviour with both sync/async mode of consumption and what does calling `storeMessage` async mean.

Some pieces are not trivial to make sense of, for example, why is it `Offset: kafka.Offset(msg.Offset + 1)` and not `Offset: kafka.Offset(msg.Offset)`?

File: xkafka/middleware/go.mod [hunk 4]
```diff
@@ -1,6 +1,8 @@
 module github.com/gojekfarm/xtools/xkafka/middleware
 
-go 1.19
+go 1.21
+
+toolchain go1.21.0
```
[the remark this answers is not in the document]
Comment (sonnes) [reply xkafka/middleware/go.mod:5]: https://go.dev/blog/toolchain

File: xkafka/consumer.go (hunk 2 above)
Comment (sonnes) [reply xkafka/consumer.go:212]: I think we can expose the `Commit` method to allow have their own commit strategy.

So far I have not seen too many use cases of manual commit, hence keeping it simple for now.

File: xkafka/consumer.go (hunk 3 above)
Comment (sonnes) [reply xkafka/consumer.go:201]: > why is it Offset: kafka.Offset(msg.Offset + 1) and not Offset: kafka.Offset(msg.Offset)?

This is based on https://github.com/confluentinc/confluent-kafka-go/blob/af4a5f8b2018db6503f7e8097a25a24a6d6feb06/kafka/consumer.go#L374

it ensures that the consumer starts with next message when it restarts

Comment (ajatprabha) [reply xkafka/consumer.go:201]: Okay, let's add this as a doc comment at relevant line in code for other readers.

File: xkafka/consumer.go [hunk 5]
```diff
@@ -77,57 +84,135 @@ func (c *Consumer) Start(ctx context.Context) error {
...
+			return err
+		default:
+			km, err := c.kafka.ReadMessage(c.config.pollTimeout)
+			if err != nil {
+				var kerr kafka.Error
+				if errors.As(err, &kerr) && kerr.Code() == kafka.ErrTimedOut {
+					continue
+				}
+
+				if ferr := c.config.errorHandler(err); ferr != nil {
```
Comment (ajatprabha) [inline xkafka/consumer.go:113]: When `errorHandler` return `nil` from here. The iteration is not skipped, and it invokes `newMessage` causing nil deref, ideally in-case handler returns nil at this point, iteration should be skipped (with `continue`).

Comment (ajatprabha) [reply xkafka/consumer.go:113]: Example:
```go
xkafka.ErrorHandler(func(err error) error {
	var kerr kafka.Error
	if errors.As(err, &kerr) && kerr.Code() == kafka.ErrUnknownTopicOrPart {
		return nil
	}

	return err
})
```

Or 
```go
xkafka.ErrorHandler(func(err error) error {
	var kerr kafka.Error
	if errors.As(err, &kerr) && !kerr.IsFatal() {
		return nil
	}

	return err
})
```

File: xkafka/consumer.go [hunk 6]
```diff
@@ -77,57 +84,135 @@ func (c *Consumer) Start(ctx context.Context) error {
...
 			km, err := c.kafka.ReadMessage(c.config.pollTimeout)
-
 			if err != nil {
-				if kerr, ok := err.(kafka.Error); ok && kerr.Code() == kafka.ErrTimedOut {
+				var kerr kafka.Error
+				if errors.As(err, &kerr) && kerr.Code() == kafka.ErrTimedOut {
 					continue
 				}
 
 				if ferr := c.config.errorHandler(err); ferr != nil {
```
Comment (ajatprabha) [inline xkafka/consumer.go:162]: Similar to [comment](https://github.com/gojekfarm/xtools/pull/26/files#r1428501491)

```go
if ferr := c.config.errorHandler(err); ferr != nil {
	cancel(ferr)
}

continue
```

**Review by ajatprabha — APPROVED** [review]

**Pull request merged**